### PR TITLE
test/recipes/test_genrsa.t : don't fail because of size limit changes

### LINKED ISSUE #5751: Regression test failures, 1.1.0 test programs vs 1.1.1 (master) library [2018-03-26]
Compare to last time I did this, it looks better, much better...

```
# OPENSSL_REGRESSION gets the absolute path to my 'master' build
: ; OPENSSL_REGRESSION=$(cd ..; pwd)/_build make -j8 test 
make depend && make _tests
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_build-1.1.0'
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_build-1.1.0'
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_build-1.1.0'
( cd test; \
  mkdir -p test-runs; \
  SRCTOP=../../1.1.0 \
  BLDTOP=../. \
  RESULT_D=test-runs \
  PERL="/usr/bin/perl" \
  EXE_EXT= \
  OPENSSL_ENGINES=`cd .././engines; pwd` \
  OPENSSL_DEBUG_MEMORY=on \
    /usr/bin/perl ../../1.1.0/test/run_tests.pl  )
../../1.1.0/test/recipes/01-test_abort.t ............ ok   
../../1.1.0/test/recipes/01-test_sanity.t ........... ok   
../../1.1.0/test/recipes/01-test_symbol_presence.t .. ok   
../../1.1.0/test/recipes/02-test_ordinals.t ......... ok   
../../1.1.0/test/recipes/03-test_exdata.t ........... ok   
../../1.1.0/test/recipes/03-test_ui.t ............... ok   
../../1.1.0/test/recipes/04-test_pem.t .............. ok     
../../1.1.0/test/recipes/05-test_bf.t ............... ok   
../../1.1.0/test/recipes/05-test_cast.t ............. ok   
../../1.1.0/test/recipes/05-test_des.t .............. ok   
../../1.1.0/test/recipes/05-test_hmac.t ............. ok   
../../1.1.0/test/recipes/05-test_idea.t ............. ok   
../../1.1.0/test/recipes/05-test_md2.t .............. skipped: md2 is not supported by this OpenSSL build
../../1.1.0/test/recipes/05-test_md4.t .............. ok   
../../1.1.0/test/recipes/05-test_md5.t .............. ok   
../../1.1.0/test/recipes/05-test_mdc2.t ............. ok   
../../1.1.0/test/recipes/05-test_rand.t ............. ok   
../../1.1.0/test/recipes/05-test_rc2.t .............. ok   
../../1.1.0/test/recipes/05-test_rc4.t .............. ok   
../../1.1.0/test/recipes/05-test_rc5.t .............. skipped: rc5 is not supported by this OpenSSL build
../../1.1.0/test/recipes/05-test_rmd.t .............. ok   
../../1.1.0/test/recipes/05-test_sha1.t ............. ok   
../../1.1.0/test/recipes/05-test_sha256.t ........... ok   
../../1.1.0/test/recipes/05-test_sha512.t ........... ok   
../../1.1.0/test/recipes/05-test_wp.t ............... ok   
../../1.1.0/test/recipes/10-test_bn.t ............... ok   
../../1.1.0/test/recipes/10-test_exp.t .............. ok   
../../1.1.0/test/recipes/15-test_dh.t ............... ok   
../../1.1.0/test/recipes/15-test_dsa.t .............. ok   
../../1.1.0/test/recipes/15-test_ec.t ............... ok   
../../1.1.0/test/recipes/15-test_ecdsa.t ............ ok   
../../1.1.0/test/recipes/15-test_genrsa.t ........... Dubious, test returned 4 (wstat 1024, 0x400)
Failed 4/5 subtests 
../../1.1.0/test/recipes/15-test_rsa.t .............. ok   
../../1.1.0/test/recipes/15-test_rsapss.t ........... ok   
../../1.1.0/test/recipes/20-test_enc.t .............. ok       
../../1.1.0/test/recipes/20-test_passwd.t ........... ok   
../../1.1.0/test/recipes/25-test_crl.t .............. ok   
../../1.1.0/test/recipes/25-test_d2i.t .............. ok     
../../1.1.0/test/recipes/25-test_pkcs7.t ............ ok   
../../1.1.0/test/recipes/25-test_req.t .............. ok   
../../1.1.0/test/recipes/25-test_sid.t .............. ok   
../../1.1.0/test/recipes/25-test_verify.t ........... ok       
../../1.1.0/test/recipes/25-test_x509.t ............. ok   
../../1.1.0/test/recipes/30-test_afalg.t ............ ok   
../../1.1.0/test/recipes/30-test_engine.t ........... ok   
../../1.1.0/test/recipes/30-test_evp.t .............. ok   
../../1.1.0/test/recipes/30-test_evp_extra.t ........ ok   
../../1.1.0/test/recipes/30-test_pbelu.t ............ ok   
../../1.1.0/test/recipes/40-test_rehash.t ........... ok   
../../1.1.0/test/recipes/60-test_x509_store.t ....... ok   
../../1.1.0/test/recipes/70-test_asyncio.t .......... ok   
../../1.1.0/test/recipes/70-test_bad_dtls.t ......... ok   
../../1.1.0/test/recipes/70-test_clienthello.t ...... ok   
../../1.1.0/test/recipes/70-test_packet.t ........... ok   
../../1.1.0/test/recipes/70-test_sslcbcpadding.t .... ok   
../../1.1.0/test/recipes/70-test_sslcertstatus.t .... ok   
../../1.1.0/test/recipes/70-test_sslextension.t ..... Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/3 subtests 
../../1.1.0/test/recipes/70-test_sslmessages.t ...... ok   
../../1.1.0/test/recipes/70-test_sslrecords.t ....... Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/11 subtests 
../../1.1.0/test/recipes/70-test_sslsessiontick.t ... ok    
../../1.1.0/test/recipes/70-test_sslskewith0p.t ..... ok   
../../1.1.0/test/recipes/70-test_sslvertol.t ........ ok   
../../1.1.0/test/recipes/70-test_tlsextms.t ......... ok   
../../1.1.0/test/recipes/70-test_verify_extra.t ..... ok   
../../1.1.0/test/recipes/80-test_ca.t ............... ok   
../../1.1.0/test/recipes/80-test_cipherlist.t ....... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
../../1.1.0/test/recipes/80-test_cms.t .............. ok   
../../1.1.0/test/recipes/80-test_ct.t ............... ok   
../../1.1.0/test/recipes/80-test_dane.t ............. ok   
../../1.1.0/test/recipes/80-test_dtls.t ............. ok   
../../1.1.0/test/recipes/80-test_dtlsv1listen.t ..... ok   
../../1.1.0/test/recipes/80-test_ocsp.t ............. ok     
../../1.1.0/test/recipes/80-test_pkcs12.t ........... ok   
../../1.1.0/test/recipes/80-test_ssl_new.t .......... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/19 subtests 
../../1.1.0/test/recipes/80-test_ssl_old.t .......... ok   
../../1.1.0/test/recipes/80-test_ssl_test_ctx.t ..... ok   
../../1.1.0/test/recipes/80-test_sslcorrupt.t ....... ok   
../../1.1.0/test/recipes/80-test_tsa.t .............. ok     
../../1.1.0/test/recipes/80-test_x509aux.t .......... ok   
../../1.1.0/test/recipes/90-test_async.t ............ ok   
../../1.1.0/test/recipes/90-test_bio_enc.t .......... ok   
../../1.1.0/test/recipes/90-test_bioprint.t ......... ok   
../../1.1.0/test/recipes/90-test_constant_time.t .... ok   
../../1.1.0/test/recipes/90-test_fatalerr.t ......... ok   
../../1.1.0/test/recipes/90-test_fuzz.t ............. ok     
../../1.1.0/test/recipes/90-test_gmdiff.t ........... ok   
../../1.1.0/test/recipes/90-test_heartbeat.t ........ skipped: heartbeats is not supported by this OpenSSL build
../../1.1.0/test/recipes/90-test_ige.t .............. ok   
../../1.1.0/test/recipes/90-test_memleak.t .......... ok   
../../1.1.0/test/recipes/90-test_p5_crpt2.t ......... ok   
../../1.1.0/test/recipes/90-test_secmem.t ........... ok   
../../1.1.0/test/recipes/90-test_shlibload.t ........ ok   
../../1.1.0/test/recipes/90-test_srp.t .............. ok   
../../1.1.0/test/recipes/90-test_sslapi.t ........... ok   
../../1.1.0/test/recipes/90-test_threads.t .......... ok   
../../1.1.0/test/recipes/90-test_v3name.t ........... ok   

Test Summary Report
-------------------
../../1.1.0/test/recipes/15-test_genrsa.t         (Wstat: 1024 Tests: 5 Failed: 4)
  Failed tests:  2-5
  Non-zero exit status: 4
../../1.1.0/test/recipes/70-test_sslextension.t   (Wstat: 512 Tests: 3 Failed: 2)
  Failed tests:  2-3
  Non-zero exit status: 2
../../1.1.0/test/recipes/70-test_sslrecords.t     (Wstat: 512 Tests: 11 Failed: 2)
  Failed tests:  4, 6
  Non-zero exit status: 2
../../1.1.0/test/recipes/80-test_cipherlist.t     (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
../../1.1.0/test/recipes/80-test_ssl_new.t        (Wstat: 256 Tests: 19 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
Files=96, Tests=556, 199 wallclock secs ( 1.78 usr  0.17 sys + 132.02 cusr 59.00 csys = 192.97 CPU)
Result: FAIL
make[1]: *** [Makefile:158: _tests] Error 1
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_build-1.1.0'
make: *** [Makefile:155: tests] Fel 2
```

### PR DESCRIPTION
**Pull request by levitte**
There is a test to check that 'genrsa' doesn't accept absurdly low
number of bits.  Apart from that, this test is designed to check the
working functionality of 'openssl genrsa', so instead of having a hard
coded lower limit on the size key, let's figure out what it is.

Partially fixes #5751

### CODE REVIEW

**Comment by richsalz** [comment]
This is not worth fixing until we have addressed the 1.1.1 PR's.

File: test/recipes/15-test_genrsa.t [hunk 1]
```diff
@@ -18,9 +18,34 @@ setup("test_genrsa");
...
-ok(run(app([ 'openssl', 'genrsa', '-3', '-out', 'genrsatest.pem', '16'])), "genrsa -3 16");
-ok(run(app([ 'openssl', 'rsa', '-check', '-in', 'genrsatest.pem', '-noout'])), "rsa -check");
-ok(run(app([ 'openssl', 'genrsa', '-f4', '-out', 'genrsatest.pem', '16'])), "genrsa -f4 16");
-ok(run(app([ 'openssl', 'rsa', '-check', '-in', 'genrsatest.pem', '-noout'])), "rsa -check");
+
+# Depending on the shared library, we might have different lower limits.
+# Let's find it!
+note "Looking for lowest amount of bits";
+my $bad = 3;                    # Number of bits
+my $good = 11;                  # Number of bits
```
Comment (dot-asm) [inline test/recipes/15-test_genrsa.t]: No, they are not. I mean these are not amount of bits, but binary logarithm of amount of bits, or number of bits in number holding number of [good] bits.

However! I'd rather suggest to simply take the recipe from master to 1.1.0. It's really no point in testing such low numbers.

**Comment by dot-asm** [comment]
> This is not worth fixing until we have addressed the 1.1.1 PR's.

I disagree, because this kind of cross-release tests can indicate problems in 1.1.1, and sooner we get it pass is getter.

File: test/recipes/15-test_genrsa.t (hunk 1 above)
Comment (levitte) [reply test/recipes/15-test_genrsa.t]: The script from master will fail because of this line:

```
is(run(app([ 'openssl', 'genrsa', '-3', '-out', 'genrsatest.pem', '256'])), 0, "genrsa -3 256");
```

Of course, that particular line can be left with `'8'` rather than `'256'`.  It *is* designed to check that we do have a lower bound, even though absurd.

I'm also thinking that some time in the future, we'll end up with the exact same problem again, just with higher numbers.  Someone will get tired of the repeating scenario...

Re comment on number of bits, of course you're right.

**Comment by levitte** [comment]
@richsalz, this is about removing false positives in a regression failure test.  I hope that avoiding regression failures, and being able to show that we don't have them, at least to ourselves, is something we can agree should be done.  If you have a different opinion, I think it should be taken to the openssl-project list.

**Comment by richsalz** [comment]
I didn't say it's not worth doing.  I said fixing 1.1.0 issues is less important than working on our 1.1.1 exit criteria.

**Comment by levitte** [comment]
Errrr, you do realise that things related to #5751 *are* about 1.1.1, right?

**Comment by richsalz** [comment]
I misunderstood the importance and relevance, sorry for the distraction and thanks for the clarification.

**Comment by levitte** [comment]
Thanks, and you're welcome :-)

File: test/recipes/15-test_genrsa.t (hunk 1 above)
Comment (dot-asm) [reply test/recipes/15-test_genrsa.t]: As for future, this won't cross 1.1.x boundary, so it will be just limited number of deviations. And if one puts [that much] value into checking the lower boundary, it can be the only difference between the branches. I mean one always takes recipe from master and adjusts down only the first test for lower bound.

Concern with iterative search is that algorithm is hard to follow... And formally speaking it might need adjustment in new version anyway. Because choice of constants is formally version specific. I mean there is a cap...

Comment (dot-asm) [reply test/recipes/15-test_genrsa.t]: > there is a cap

For example if you set minimum to any value over 2048 it will fail to find right value. I'm not saying that we should set it to 2048, only that suggested algorithm has limitations (that one has to keep in mind and take into consideration). I'd say that examining openssl_versions output (introduced in #5753) would be more appropriate. Especially with commentary that says that limit modifications has to be backported to earlier minor versions.

Comment (levitte) [reply test/recipes/15-test_genrsa.t]: Either way, there will be something to maintain at some point, it seems.  I fail to understand why one is superior to the other...

File: test/recipes/15-test_genrsa.t [hunk 2]
```diff
@@ -18,9 +18,34 @@ setup("test_genrsa");
...
-ok(run(app([ 'openssl', 'rsa', '-check', '-in', 'genrsatest.pem', '-noout'])), "rsa -check");
-ok(run(app([ 'openssl', 'genrsa', '-f4', '-out', 'genrsatest.pem', '16'])), "genrsa -f4 16");
-ok(run(app([ 'openssl', 'rsa', '-check', '-in', 'genrsatest.pem', '-noout'])), "rsa -check");
+
+# Depending on the shared library, we might have different lower limits.
+# Let's find it!
+note "Looking for lowest amount of bits";
+my $bad = 3;                    # Log2 of number of bits
+my $good = 11;                  # Log2 of number of bits
+my $checked = int(($good + $bad + 1) / 2);
```
Comment (paulidale) [inline test/recipes/15-test_genrsa.t]: Wouldn't this be better at the start inside the loop, avoiding the same expression twice?

**Review by paulidale — APPROVED** [review]
Approved with or without the suggested change.

The binary search seems like overkill but it is future proof.

File: test/recipes/15-test_genrsa.t (hunk 1 above)
Comment (dot-asm) [reply test/recipes/15-test_genrsa.t]: > I fail to understand why one is superior to the other...

As already said, suggested algorithm is hard to follow, and consequently hard to maintain. As practical example, if one wants to increase the cap beyond 2048, how many minutes will it take for somebody else to figure out how to do it and at the same time ensure that it still work in 1.1.0 context. Compare it to adding *one* number, minimum supported, to lookup table with version as "index".

**Comment by levitte** [comment]
Added some commentary that might be helpful

**Comment by levitte** [comment]
Merged.

master / 1.1.1:
1b9f41a0df6ed2fac1c5303be909c2f70c404f20 test/recipes/test_genrsa.t : don't fail because of size limit changes

1.1.0:
ec46830f8a4ce62c0c8ee7677b1eb8e53ee16df1 test/recipes/test_genrsa.t : don't fail because of size limit changes
